### DISCUSSION TITLE
Does it support Line Obstacle?

### DISCUSSION THREAD

**Discussion by sunsvip** [Nebukam/com.nebukam.orca#7] [Q&A] [answered]
Hello, thank you for freely sharing such a great project. Compared to the regular multithreaded RVO2, the Job System has significantly improved. However, the original RVO2 supports adding a line obstacle using two points, which the ORCA implementation cannot achieve. Is there a bug in the ObstacleOrientationJob.cs script? How can this be fixed? Thank you!

**Comment by Nebukam** [comment]
Hey there @sunsvip ! Short answer: **YES**.

ORCA obstacles are edge-based (or segment-based if you will). Any obstacle is really just a polygon, and it can be an open one, so a RVO2 LineObstacle is really just a single-edge polygon with ORCA.

You can test it for yourself with the provided Setup example, by changing the Min/Max obstacle edge with a value of 2, which effectively generates a single edge:  

![image](https://github.com/Nebukam/com.nebukam.orca/assets/5935407/fb76a858-822f-4df7-a45c-6e590c1de701)
![image](https://github.com/Nebukam/com.nebukam.orca/assets/5935407/0b3cded2-574b-47b1-9969-3f0e3d36b1ab)

Roughly the syntax to add an edge would look like this:
```CSharp
ObstacleGroup obstacles = new ObstacleGroup(); //Obstacle group used with the simulation

float3 pointA, pointB;
float3[] edge = { pointA, pointB }; //An edge made of two points

obstacles.Add(edge);
```

Hope this helps!

**Reply by sunsvip** [reply]
Thank you for your response. Here is my code：
![test3](https://github.com/Nebukam/com.nebukam.orca/assets/16739213/43f8a1c2-9d35-4c99-a6dd-925472f678bd)
`protected override void Awake()
    {
        base.Awake();
        m_orca = new ORCA();
        m_agents = new AgentGroup<Agent>();
        m_staticObstacles = new ObstacleGroup();
    }

    private void Start()
    {
        m_orca.plane = m_plane;// XZ
        m_orca.agents = m_agents;
        m_orca.staticObstacles = m_staticObstacles;

        //Add edge obstacles
        var edgePoints = new float3[]
        {
            float3(50, 0, 50),
            float3(50,0,-50),
            float3(-50,0,-50),
            float3(-50, 0, 50),
            float3(50, 0, 50)
        };

        for (int i = 0; i < edgePoints.Length - 1; i++)
        {
            m_staticObstacles.Add(new float3[] { edgePoints[i], edgePoints[i + 1] });
        }
    }`

**Reply by Nebukam** [reply]
Aaaah yes I see what's going on here. Edges collision are one-sided based on their orientation (first vertex -> second vertex).

So you go three options here:

### **Option A**
Based on your specific use-case, a square, simply make a single square collision polygon instead of four individual obstacles edges like so:
```CSharp
var edgePoints = new float3[]
    {
        float3(50, 0, 50), //A
        float3(50,0,-50),  //B
        float3(-50,0,-50), //C
        float3(-50, 0, 50), //D
        float3(50, 0, 50) //A
    }; // A -> B -> C -> D -> wrap to A

 m_staticObstacles.Add(edgePoints); // Just use your array as-is :D
```
*(It's worth noting that `ObstacleGroup.Add` takes a second parameter, which is whether or not to reverse the input vertices order of the obstacle)*
If the collision is "reversed" change it to `m_staticObstacles.Add(edgePoints, true);` or manually change the order of points within the array 

![image](https://github.com/Nebukam/com.nebukam.orca/assets/5935407/4c65b700-c6ef-4cc3-97f4-c2bbed4ab317)


### **Option B:**
Althought an odd approach, if you really need floating edges you could create two obstacles with opposite orientations like so:
```CSharp
var edge = new float3[] { pointA, pointB };
 m_staticObstacles.Add(edge , false); // A -> B
 m_staticObstacles.Add(edge, true ); // B -> A
```

### **Option C:**
Create an acute 3-points polyline like so (but you may end up with missed maths from time to time):
```CSharp
var edge = new float3[] { pointA, pointB, pointA };
 m_staticObstacles.Add(edge , false); // A -> B -> A
```

**Reply by sunsvip** [reply]
Thank you for your reply. Yes, I've tried these approaches and they all failed. I seem to have found some clues: for polygonal obstacles, the vertex order must be counterclockwise. For the bounding box, the vertex order must be clockwise. However, there still exists a problem, as shown in the screenshot below: obs 0 and obs 1 are two identical Edge Obstacles. Obs 0 does not contain any other obstacles inside, so it can successfully block the agent within its range. However, obs 1 contains an obstacle inside, which seems to prevent it from effectively blocking the agent completely.
![test3](https://github.com/Nebukam/com.nebukam.orca/assets/16739213/55b1ff3a-540a-4485-8620-b79bdd570fb7)
![test4](https://github.com/Nebukam/com.nebukam.orca/assets/16739213/0393ebb4-e032-4d17-a2f4-1953099b0dcd)
![image](https://github.com/Nebukam/com.nebukam.orca/assets/16739213/92010b0b-d240-45e9-b935-e2afe99eca8c)

**Reply by sunsvip** [reply]
When I remove the obstacles inside Edge Obstacle, the agent no longer overflows the boundaries.
![test4](https://github.com/Nebukam/com.nebukam.orca/assets/16739213/e239b408-8737-4f6d-ac6c-a225ef23404a)

**Reply by Nebukam** [reply]
Hmmm I'll be looking into this, I've been able to reproduce the issue on my end.
I suspect the maths and the way I'm building the KD tree for obstacles fails to recognize the edges as an obstacles if the vertices are too far apart if there are smaller obstacles to check.

I'll see if I can find a quick fix, althought I suspect a workaround would be to split long edges into smaller ones, i.e turning this
`x-----------------x` into this `x-----x-----x-----x`. 

If it turns out it solves the issue I'll add a some code to automate the splitting on ORCA's side and push an update, I'll keep you posted!

**Reply by Nebukam** [reply, marked as answer]
Ok so it's fixed. I pushed an update to the repo and bumped the version, so please update the package to **1.1.7** from the package manager and let me know how it works for you :)  
_(Also just in case, make sure you have Burst compile enabled in that wasn't the case before)_

#### **The problem:**
 when obstacles segments are too far apart, the agent is "blind" to these segments, leading to inconsistent behavior depending on that agent' position toward the obstacle and its vertices. 

#### **The fix:** 
Admittedly not the most elegant fix, but it works 100% of the time: 
Subdivide obstacles edges to smaller segments.
I added methods & parameters to that effect so you can keep obstacle admin simple (i.e, a square is still 4 vertices when added)
I also updated the ORCA Setup example to account for the update.

```CSharp
float3[] squarePoints = new float3[] {
    float3(min.x, min.y, 0f) * 1.2f,
    float3(min.x, max.y, 0f)* 1.2f,
    float3(max.x, max.y, 0f)* 1.2f,
    float3(max.x, min.y, 0f)* 1.2f,
};

obstacles.Add(squarePoints, false, 10.0f);
```
The last parameter of the Add method is the "maximum segment length in a edge". The smaller the value the more you are guaranteed for the obstacle to be picked up by the agents. 10.0f seems to work fine most of the time with zero hit on performance.

Alternatively if you create standalone Obstacle objects  (i.e without the shorthand on ObsctacleGroup) you can call `Obstacle.Subvidide(10.0f)` to achieve the same functionality.

Result:
![image](https://github.com/Nebukam/com.nebukam.orca/assets/5935407/c0afc789-ce86-4462-bdc2-76f2d892f28d)

Highlighted in magenta here are the final vertices pushed to the simulation:
![image](https://github.com/Nebukam/com.nebukam.orca/assets/5935407/ab648c23-ab47-42d3-aefa-267ad870b203)

**Reply by sunsvip** [reply]
Thank you very much, it works perfectly!

**Reply by Nebukam** [reply]
Thank you for taking the time to report it and be willing to try things! <3
